File: 0583-delete-operation-for-two-strings/0583-delete-operation-for-two-strings.cpp
class Solution {
public:
int t[501][501];
int check( int i,int j,string word1,string word2){

if(i>word1.size()-1){
    return word2.size()-j;
}

if(t[i][j]!=-1){
    return t[i][j];
}
if(j>word2.size()-1){
    return word1.size()-i;
}

if(word1[i]==word2[j]){
    return  t[i][j]=check(i+1,j+1,word1,word2);
}
else{

    return  t[i][j]=1+min(check(i+1,j,word1,word2),check(i,j+1,word1,word2));
}



}
    int minDistance(string word1, string word2) {
memset(t,-1,sizeof(t));
        return check(0,0,word1,word2);
        
    }
};

File: 0583-delete-operation-for-two-strings/0583-delete-operation-for-two-strings.java
class Solution {

int check(int i,int j,String word1, String word2,int[][] dp){

    int m=word1.length();
    int n=word2.length();


      if (i == m) {
            return n - j;
        }

        // word2 khatam
        if (j == n) {
            return m - i;
        }

   if(dp[i][j]!=-1){
    return dp[i][j];
}
      if (word1.charAt(i) == word2.charAt(j)) {
            return check(i + 1, j + 1, word1, word2,dp);
        }
    int a=1+check(i+1,j,word1,word2,dp);
    int b=1+check(i,j+1,word1,word2,dp);
     
    

return dp[i][j]= Math.min(a,b);
}
    public int minDistance(String word1, String word2) {
     
 int m=word1.length();
    int n=word2.length();
    int[][] dp=new int[m][n];

    for(int i=0;i<m;i++){
        Arrays.fill(dp[i],-1);
    }
        return check(0,0,word1,word2,dp);
    }
}